### ISSUE TITLE
Unexpected Behavior using Image.Load(Stream) with AWS S3 Source

### ISSUE THREAD

**Issue by ambroselittle** [SixLabors/ImageSharp#477]
### Description
I am using the .NET AWS SDK to load from S3 using their GetObject, but I had to do "extra" work to get it to work with ImageSharp, while System.Drawing.Image.FromStream worked without that extra work.

### Steps to Reproduce

1. I added a package dep via nuget to the current latest (1.0.0-beta0002).
2. Using v3.3.17 of AWS S3 SDK, I call their GetObject method and attempt to load their ResponseStream into the ImageSharp Image.Load(Stream) method. 

Actual (First Stop):
It says it can't read. I noticed AWS had their own internal "[MD5Stream](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Util/HashStream.cs#L393)" so I figured maybe I should copy that to a regular MemoryStream.

New Problem:
Can't load, tried these N formats. (All the default ones..) as in #164. That issue suggests seeking back to the beginning, which I do. And now it works.

However, that issue is closed, saying it was fixed in an alpha, but the issue persists in beta.

More to the point, and the reason I opened a separate issue, is that I don't think I should need to copy to a MemoryStream. It seems like the library should handle that, especially since "it just works" with System.Drawing. 

Basically, can the library handle all this and make it "just work"?

### System Configuration
- ImageSharp version: 1.0.0-beta0002
- Environment (Operating system, version and so on): Win10 x64 
- .NET Framework version: 4.5

**Comment by JimBobSquarePants** [comment]
Hi @ambroselittle 

>However, that issue is closed, saying it was fixed in an alpha, but the issue persists in beta.

We chose to not automatically seek to the beginning of the stream for consumers because we wanted to be able to read streams where the image might not be at the beginning of the stream (This happens).

That behaviour catering for the minority makes it cumbersome for developers who don't want to have to set the position so should change.

**What we can do:**
I can make it so that the default behaviour of the stream loading methods resets the position of the stream to **0**. Overloads offering a `ReadOrigin` parameter can choose the current position if they require.

**What we can maybe do:**
We need to be able to seek streams and copy non-seekable streams ([`MD5Stream` is non-seekable](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Util/HashStream.cs#L188)) to a `MemoryStream` to make it seekable.

You can see that here:
https://github.com/SixLabors/ImageSharp/blob/47a8533b7d7b43b490d4958d2702490a580d9d34/src/ImageSharp/Image/Image.FromStream.cs#L253

Now that relies on the ability of a stream to support `Read` which your stream doesn't. However, we may be able to simply pass the stream to the `MemoryStream` contructor instead of copying.

Is that how you wrapped the `MD5Stream`? If you can show me how you did it I can update the open PR and make it automagically work

Cheers

James

**Comment by ambroselittle** [comment]
Hi @JimBobSquarePants,

Thanks for the help.

The code looks like this:
```C#
using(var iResponse = s3.GetObject(request))
{
    fileMetaData.Stream = new MemoryStream(); // fileMetaData is our own class
    // this works -> var img = System.Drawing.Image.FromStream(iResponse.ResponseStream);
    iResponse.ResponseStream.CopyTo(fileMetaData.Stream); // CopyTo supposedly does not change seek loc
    fileMetaData.Stream.Seek(0, SeekOrigin.Begin);
   ...
}
```

**Comment by antonfirsov** [comment]
@JimBobSquarePants based on this answer, I don't think auto seeking is what the "just work" behavior meant to do here.
Looks like we just fail to copy the contents of the `MD5Stream ` to `MemoryStream` internally for some reason.

**Comment by JimBobSquarePants** [comment]
@antonfirsov 

OK.... This is getting a little confusing now. I wrote the following test code to consume an `MD5Stream` and run it past our copying code. It all seems to just work. 

@ambroselittle Have you tried the nightlies from MyGet? I'm wondering whether we have changed something since the beta? And, I know this is a daft question but the response stream is definitely an `MD5Stream`?

Is there something obvious I've missed?

``` c#
using System;
using System.IO;
using System.Security.Cryptography;
using Amazon.Runtime.Internal.Util;

namespace StreamTester
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Testing Md5Stream");

            using (var fs = File.OpenRead("Calliphora.jpg"))
            {
                var length = fs.Length;
                var hash = GetHash(fs);

                MD5Stream stream = new MD5Stream(fs, hash, length);

                Console.WriteLine($"Can read - {stream.CanRead}"); // True
                Console.WriteLine($"Can seek - {stream.CanSeek}"); // False

                Console.WriteLine($"Stream was successfully copied to a memory stream and reset to beginning - {WithSeekableStream(stream, s => s is MemoryStream && s.Position == 0)}");

                // This all prints
                // Testing Md5Stream
                // Can read - True
                // Can Seek - False
                // Stream was successfully copied to a memory stream and reset to beginning - True
            }

            Console.ReadLine();
        }

        static byte[] GetHash(FileStream fs)
        {
            using (var ms = new MemoryStream())
            {
                fs.CopyTo(ms);
                fs.Position = 0;

                using (var hashAlgorithm = MD5.Create())
                {
                    return hashAlgorithm.ComputeHash(ms.ToArray());
                }
            }
        }

        private static T WithSeekableStream<T>(Stream stream, Func<Stream, T> action)
        {
            if (!stream.CanRead)
            {
                throw new NotSupportedException("Cannot read from the stream.");
            }

            if (stream.CanSeek)
            {
                return action(stream);
            }

            // We want to be able to load images from things like HttpContext.Request.Body
            using (var memoryStream = new MemoryStream())
            {
                stream.CopyTo(memoryStream);
                memoryStream.Position = 0;
                return action(memoryStream);
            }
        }
    }
}

```

**Comment by ambroselittle** [comment]
Doh. My bad. Yup, definitely missing context that would prevent the repro. In my effort to simplify the code sample..

Bottom line is my `using` was in a method that assigned/returned a reference to the `MD5Stream`. I'm assuming that `Dispose` must mark the stream as unreadable, which is odd considering that copying it, it was still actually readable--even though I have to reset the position (which is weird given that the docs say `CopyTo` shouldn't affect that).

I just eliminated that `using` in the method, and it can read automagically as expected/desired. Sorry about that. User error. Many thanks for your alacrity on the issue.

**Comment by JimBobSquarePants** [comment]
@ambroselittle I'm just glad we figured this out! 😄 

I don't know why System.Drawing would have worked on a disposed stream, maybe they don't do our additional check. We don't want to remove that `CanRead` check though as it eliminates potential bugs.

I think this discussion has proved fruitful though and the addition of reading control for seekable streams is a good one.

I'll close this once the associated PR is merged.

**Comment by ambroselittle** [comment]
I think because it was inline there where you see it in the sample, i.e., inside the using. I tossed that in there without thinking about the different context myself, though it's obvious to me now..

**Comment by JimBobSquarePants** [comment]
Ah of course!
